src/v/compression/stream_zstd.cc

```diff
@@ -46,18 +67,16 @@ stream_zstd::zstd_compress_ctx& stream_zstd::compressor() {
     return _compress;
 }
 
-void stream_zstd::reset_decompressor() {
-    _decompress.reset(ZSTD_createDCtx());
-    if (!_decompress) {
-        throw std::bad_alloc{};
-    }
-}
-
-stream_zstd::zstd_decompress_ctx& stream_zstd::decompressor() {
-    if (!_decompress) {
-        reset_decompressor();
+ZSTD_DCtx* stream_zstd::decompressor() {
+    if (unlikely(!dctx_workspace)) {
+        /*
+         * handle on-demand case for tests, etc..
+         */
+        init_workspace(2_MiB);
     }
-    return _decompress;
+    auto ctx = ZSTD_initStaticDCtx(dctx_workspace.get(), dctx_workspace_size);
+    vassert(ctx, "Could not initialize static decompression context");
```

Comment:
isn't this check implicitly checked in the prev `init_workspace` ? i.e.: can this fail?

Reply:
no, this is the initialization of the context and init_workspace only does an allocation